### ISSUE TITLE
[ENH]: Provide a way to avoid subcommands on import.

### ISSUE THREAD

**Issue by craigcitro-ant** [matplotlib/matplotlib#28488]
### Problem

tl;dr: I'd like to be able to avoid `from matplotlib import pyplot as plt` invoking subprocesses at import time.

More generally: we often run jobs which involve a transitive import of matplotlib; for reasons I do not fully understand, we see rare but regular flakes where the process hangs during import, specifically [at this line](https://github.com/matplotlib/matplotlib/blob/74c7f9a598c4e32b3f551f75ed965f446d786ab1/lib/matplotlib/font_manager.py#L256) where matplotlib first invokes `fc-list`. The code is written to be robust to `fc-list` not existing, but here it's the existence check itself that's causing trouble.

I'd like to be able to disable that call; in general, I suspect "this library is running subcommands during package import" may well cause these sorts of problems in other environments, so I suspect this would be helpful to other users as well.

### Proposed solution

I've got a quick first pass at the type of change I'm describing at https://github.com/craigcitro-ant/matplotlib/commit/46d64f26977c5a37074ece74d602051da2df09e7; if that looks acceptable, I'm more than happy to clean it up and send a PR. (The main thing that's missing is adding a mention to the docs.)

**Comment by tacaswell** [comment]
Related to https://github.com/matplotlib/matplotlib/issues/28485

If we do this the flag should be something like `MPL_IGNORESYSTEMFONTS`. I suspect the same thing could be applied to the macOS check as well.  

We only call the code that calls subprocesses if we do not find a cached version of the results https://github.com/matplotlib/matplotlib/blob/f5067c16757f4cd800ae3ba0ac25ff79d2e41a7f/lib/matplotlib/font_manager.py#L1567-L1585

If you ensure that cache is created (and cached) in your pipelines then you will also avoid this problem (and speed up your job startup!).  For example, if you are using docker files adding `RUN python -c "import matplotlib.font_manager"` as the last line should be enough (and we already have tools to configure where we read/write the configuration data).

**Comment by craigcitro-ant** [comment]
thanks for the quick reply!

agreed that in principle "always create the font cache before pipelines start" is a good discipline to have, but the tradeoff is that it means every image used anywhere in one of our pipelines needs to do have *already done this*: it's a build-time thing, not a runtime thing. I think there's still value in having the ability to disable this at runtime; WDYT?

re: details: agreed on disabling for both OS variants at the same time, and happy to update the env var.

**Comment by tacaswell** [comment]
On a bit more thought, I think the best path here is to add an option to ignore _all_ fonts that are not what we bundle (which will probably make headaches for at least debian who de-bundle the fonts).  

Even if we drop just the sub-process we still do some directory walking and attempt to discover fonts which is requires a bunch of filesystem work and then I/O as we read the files to extract some information which can be slow (which is why there is the threading work in there) and leaving it in a state where we find _some_ system fonts (because on some CI systems subprocesses hang(?!)) does not feel right.  On the other hand "I want to have every font accessible on my system available to Matplotlib" vs "do not waste time looking for other fonts, I know I'm going to use the defaults" seems like a reasonable decision point.  The first is good for interactive use (where we assume that the users home directory is presistent and fonts should "just work") and the second is good for CI / automated jobs where the home directory is not persistent. 

On the other hand, this does come at the cost of not (automatically) providing access to non-western fonts (e.g. we do not bundle a font that includes any CJK gylphs).

**Comment by dvgica** [comment]
I added a comment to a different, but related, issue here: https://github.com/matplotlib/matplotlib/issues/28485#issuecomment-2209235627.

@tacaswell What you're suggesting makes sense I think. Sounds like it would remove the need for us to pre-populate the font cache at build time.
